### PR TITLE
Fix `subtitleTextStyle.color` isn't applied to the `ListTile.subtitle` in Material 2

### PR DESCRIPTION
**Pull request by burakJs**
The difference between header text style and subtitle text style and the reason why it doesn't work is the code difference below. If we make the subtitle text style the same as the title text style it will work

<details>
  <summary>Title Text Style</summary>
  
  ###  All Code
  
  ```dart
  TextStyle titleStyle = titleTextStyle
      ?? tileTheme.titleTextStyle
      ?? defaults.titleTextStyle!;
    final Color? titleColor = effectiveColor;
    titleStyle = titleStyle.copyWith(
      color: titleColor,
      fontSize: _isDenseLayout(theme, tileTheme) ? 13.0 : null,
    );
    final Widget titleText = AnimatedDefaultTextStyle(
      style: titleStyle,
      duration: kThemeChangeDuration,
      child: title ?? const SizedBox(),
    );
  ```
  
  ## Different Code Section
  
  ```dart
  final Color? titleColor = effectiveColor;
  ```
</details>

<details>
  <summary>Subtitle Text Style</summary>
  
  ## All Code
  
  ```dart
  subtitleStyle = subtitleTextStyle
        ?? tileTheme.subtitleTextStyle
        ?? defaults.subtitleTextStyle!;
      final Color? subtitleColor = effectiveColor
        ?? (theme.useMaterial3 ? null : theme.textTheme.bodySmall!.color);
      subtitleStyle = subtitleStyle.copyWith(
        color: subtitleColor,
        fontSize: _isDenseLayout(theme, tileTheme) ? 12.0 : null,
      );
      subtitleText = AnimatedDefaultTextStyle(
        style: subtitleStyle,
        duration: kThemeChangeDuration,
        child: subtitle!,
      );
  ```
  
  ## Different Code Section
  
  ```dart
  final Color? subtitleColor = effectiveColor
        ?? (theme.useMaterial3 ? null : theme.textTheme.bodySmall!.color);
  ```

### Description for code 
- The value `theme.textTheme.bodySmall!.color` is given because the `effectiveColor` value is `null` and the `theme.useMaterial3` value is `false`
</details>

<details>
  <summary>Problem solved code</summary>
  
  ## All Code
  
  ```dart
  subtitleStyle = subtitleTextStyle
        ?? tileTheme.subtitleTextStyle
        ?? defaults.subtitleTextStyle!;
      final Color? subtitleColor = effectiveColor;
      subtitleStyle = subtitleStyle.copyWith(
        color: subtitleColor,
        fontSize: _isDenseLayout(theme, tileTheme) ? 12.0 : null,
      );
      subtitleText = AnimatedDefaultTextStyle(
        style: subtitleStyle,
        duration: kThemeChangeDuration,
        child: subtitle!,
      );
  ```
</details>

<details>
<summary>Screenshot of the result after making the necessary change</summary>

</details>

#133412

*If you had to change anything in the [flutter/tests] repo, include a link to the migration guide as per the [breaking change policy].*

## Pre-launch Checklist

- [X] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [X] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [X] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [X] I signed the [CLA].
- [X] I listed at least one issue that this PR fixes in the description above.
- [X] I updated/added relevant documentation (doc comments with `///`).
- [X] I added new tests to check the change I am making, or this PR is [test-exempt].
- [X] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[test-exempt]: https://github.com/flutter/flutter/wiki/Tree-hygiene#tests
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[Features we expect every widget to implement]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat

### CODE REVIEW

**Comment by burakJs** [comment]
> It looks like there are some test failures.

I'm investigating and will let you know as soon as possible.

**Comment by burakJs** [comment]
I couldn't find a correlation between where I made code changes and tests that failed. Can you take a look please @TahaTesser

**Comment by TahaTesser** [comment]
This was effectively only updated for Material 3 without changing the M2 behavior.
 
This way can have the correct color applied in M3 without breaking existing tests and apps.
```dart
      final Color? subtitleColor = effectiveColor
        ?? (theme.useMaterial3 ? null : theme.textTheme.bodySmall!.color);
```

It looks like your change for M2 is breaking some customer tests.

```console
| 00:18 +78 ~1: /b/s/w/ir/x/t/flutter_customer_testing.flutter_packages.OTZXUG/tests/packages/rfw/test/material_widgets_test.dart: Material widgets
| ══╡ EXCEPTION CAUGHT BY FLUTTER TEST FRAMEWORK ╞════════════════════════════════════════════════════
| The following assertion was thrown while running async test code:
| Golden "goldens/material_test.drawer.png": Pixel test failed, 1.02% diff detected.
| Failure feedback can be found at
| /b/s/w/ir/x/t/flutter_customer_testing.flutter_packages.OTZXUG/tests/packages/rfw/test/failures
```

 Material 2 will eventually be deprecated and this issue will become irrelevant. 


cc: @HansMuller

**Comment by burakJs** [comment]
```dart
final Color? subtitleColor = effectiveColor
    ?? (theme.useMaterial3 ? null : theme.textTheme.bodySmall!.color);
```

- So why do we change the color property of the subtitleTextStyle given by the user in this code?

**Comment by TahaTesser** [comment]
> ```dart
> final Color? subtitleColor = effectiveColor
>     ?? (theme.useMaterial3 ? null : theme.textTheme.bodySmall!.color);
> ```
> 
> * So why do we change the color property of the subtitleTextStyle given by the user in this code?

`textColor` & `selectedColor` overrides both `titleTextStyle` & `subtitleTextStyle`, that's `effectiveColor`


If both `text` & `selectedColor` are null, we default to a `null` value so the recently added `subtitleTextStyle`'s color can be used only for M3 and we default `bodySmall` color`  for backward compatibility. 

`bodySmall!.color` has been used for the subtitle in M2 as the default color and you're changing this default behavior hence the test issues.

If we still want the `subtitleTextStyle` color fo the M2 we can do this without the changing M2 default. Please apply the I suggested and revert the default test update.

File: packages/flutter/lib/src/material/list_tile.dart
Comment (TahaTesser) [inline packages/flutter/lib/src/material/list_tile.dart:804]: Please apply this fix.
```dart
    if (subtitle != null) {
      subtitleStyle = subtitleTextStyle ?? tileTheme.subtitleTextStyle;
      if (theme.useMaterial3) {
        subtitleStyle = (subtitleStyle ?? defaults.subtitleTextStyle)!.copyWith(
          color: effectiveColor,
          fontSize: _isDenseLayout(theme, tileTheme) ? 12.0 : null,
        );
      } else {
        subtitleStyle = subtitleTextStyle?.copyWith(
          color: effectiveColor,
          fontSize: _isDenseLayout(theme, tileTheme) ? 12.0 : null,
        );
        subtitleStyle ??= defaults.subtitleTextStyle!.copyWith(
          color: effectiveColor ??  theme.textTheme.bodySmall!.color,
          fontSize: _isDenseLayout(theme, tileTheme) ? 12.0 : null,
        );
      }
      subtitleText = AnimatedDefaultTextStyle(
        style: subtitleStyle,
        duration: kThemeChangeDuration,
        child: subtitle!,
      );
    }
```

File: packages/flutter/test/material/list_tile_test.dart
Comment (TahaTesser) [inline packages/flutter/test/material/list_tile_test.dart:3619]: Please revert this default test update.
```suggestion
      expect(subtitle.text.style!.color, theme.textTheme.bodySmall!.color);
```

File: packages/flutter/test/material/list_tile_test.dart
Comment (TahaTesser) [inline packages/flutter/test/material/list_tile_test.dart:3631]:
```suggestion
      expect(subtitle.text.style!.color, theme.textTheme.bodySmall!.color);
```

**Review by TahaTesser — CHANGES_REQUESTED** [review]

**Comment by burakJs** [comment]
I will make changes and corrections as soon as possible. Thanks for the help and detailed explanation.

**Comment by burakJs** [comment]
- I made all the changes you mentioned.
- All tests in `list_tile_theme_test .dart` file pass

image

- Only Linux `web_canvaskit_tests_6` gives test timeout error

**Comment by TahaTesser** [comment]
Hey @burakJs 
Since you're a new contributor.  I've improved tests for M3 and added tests for the M2 for this PR.  
I also updated the fix itself to be much more straightforward and pushed changes myself.
I'd advise you to take a look at the changes I've made for future potential PRs.

Thank you for raising this PR.

**Review by TahaTesser — APPROVED** [review]
LGTM

**Comment by TahaTesser** [comment]
This will need a second review from @HansMuller before we land.

**Review by HansMuller — APPROVED** [review]
LGTM

**Pull request merged**
